paimon-core/src/main/java/org/apache/paimon/io/DataFileMeta.java

```diff
@@ -120,6 +121,13 @@ public class DataFileMeta {
 
     private final @Nullable List<String> valueStatsCols;
 
+    /**
+     * the data root location that the file resides in, if it is null, the file is in the default
+     * warehouse path, when {@link CoreOptions#DATA_FILE_PATH_DIRECTORY} is set, new writen files
+     * will be persisted in {@link CoreOptions#DATA_FILE_PATH_DIRECTORY}.
+     */
+    private final @Nullable String dataRootLocation;
```

Comment:
rename to externalPath?

Reply:
No problem, I'm willing to express that this path is the same root path as the warehouse, not the full path. But it doesn't affect the modification of DataFileMeta this time. I'll change it to extrenalPath first

Reply:
maybe it is better to just full path?

Reply:
I will think deeply about it in the upcoming PR, but it does not currently affect the functionality of this PR.